src/Microsoft.OData.Client/ALinq/ResourceBinder.GroupByKeySelectorAnalyzer.cs

```diff
@@ -0,0 +1,114 @@
+
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq.Expressions;
+    using System.Runtime.CompilerServices;
+
+    #endregion Namespaces
+
+    internal partial class ResourceBinder
+    {
+        /// <summary>
+        /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by.
+        /// </summary>
+        private sealed class GroupByKeySelectorAnalyzer : DataServiceALinqExpressionVisitor
+        {
+            /// <summary>The input resource, as a queryable resource.</summary>
+            private readonly QueryableResourceExpression input;
+
+            /// <summary>The key selector lambda parameter.</summary>
+            private readonly ParameterExpression lambdaParameter;
+
+            /// <summary>
+            /// Creates an <see cref="GroupByKeySelectorAnalyzer"/> expression.
+            /// </summary>
+            /// <param name="source">The source expression.</param>
+            /// <param name="paramExpr">The parameter expression.</param>
+            private GroupByKeySelectorAnalyzer(QueryableResourceExpression source, ParameterExpression paramExpr)
+            {
+                this.input = source;
+                this.lambdaParameter = paramExpr;
+            }
+
+            /// <summary>
+            /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by. 
+            /// </summary>
+            /// <param name="input">The input resource expression.</param>
+            /// <param name="keySelector">Key selector expression to analyze.</param>
+            internal static void Analyze(QueryableResourceExpression input, LambdaExpression keySelector)
+            {
+                Debug.Assert(input != null, "input != null");
+                Debug.Assert(keySelector != null, "keySelector != null");
+
+                GroupByKeySelectorAnalyzer analyzer = new GroupByKeySelectorAnalyzer(input, keySelector.Parameters[0]);
+
+                MemberInitExpression memberInitExpr = keySelector.Body as MemberInitExpression;
+
+                if (memberInitExpr != null)
+                {
+                    analyzer.Visit(memberInitExpr);
+                }
+                else
+                {
+                    analyzer.Visit(keySelector.Body);
+                }
+            }
+
+            /// <inheritdoc/>
+            internal override NewExpression VisitNew(NewExpression nex)
+            {
+                // Maintain a mapping of grouping expression and respective member
+                // The mapping is cross-referenced if any of the grouping expression 
+                // is referenced in result selector
+                for (int i = 0; i < nex.Arguments.Count; i++)
+                {
+                    this.input.Apply.GroupingExpressionsMap.Add(nex.Members[i].Name, nex.Arguments[i]);
+                }
+
+                return base.VisitNew(nex);
+            }

```

Comment:
@habbes To explain, consider the following minimal version of the `Sale`, `Customer` and `Product` models
```csharp
public class Sale
{
    public int Id { get; set; }
    public string CustomerId { get; set; }
    public Customer Customer { get; set; }
    public string ProductId { get; set; }
    public Product Product { get; set; }
}

public class Customer
{
    public string Id { get; set; }
    public string Name { get; set; }
}

public class Product
{
    public string Id { get; set; }
    public string Name { get; set; }
}
```
The key selectors when composing a `GroupBy` expression against the `Sales` entity set, could take several forms
```csharp
// Grouping by a single property
dsc.Sales.GroupBy(d1 => d1.CustomerId, /* resultSelector */)

// Grouping by a single nested property
dsc.Sales.GroupBy(d1 => d1.Customer.Name, /* resultSelector */)

// Grouping by multiple properties
dsc.Sales.GroupBy(d1 => new { d1.CustomerId, d1.ProductId }, /* resultSelector */)

// Grouping by multiple nested properties
dsc.Sales.GroupBy(d1 => new { CustomerName = d1.Customer.Name, ProductName = d1.Product.Name },
    /* resultSelector */)

```
In the scenarios where an anonymous type is used in the key selector, we create a mapping between the _class_ member (e.g. `CustomerName` in the 4th scenario) and the argument/expression assigned to it (e.g. `d1.Customer.Name`).
We maintain this mapping because that member can be used in the result selector. For example,
```csharp
dsc.Sales.GroupBy(d1 => new { CustomerName = d1.Customer.Name, ProductName = d1.Product.Name },
    (d1, d2) => new { d1.CustomerName, /* other properties in the result selector */ })
```
As we proceed with the translation and later when creating a projection plan, the mapping will serve a lookup to match the alias `CustomerName` to `d1.Customer.Name`.
This is important because the response payload won't have a property named `CustomerName`. Instead there will be something like this:
```json
{
"Customer": { "@odata.id": null, "Name": "Joe" }
}
```